src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs

```diff
@@ -0,0 +1,334 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Text.Json.Serialization.Converters;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Text.Json.Serialization.Metadata
+{
+    public partial class JsonTypeInfo<T>
+    {
+        // This section provides helper methods guiding root-level serialization
+        // of values corresponding according to the current JsonTypeInfo configuration.
+
+        // Root serialization method for sync, non-streaming serialization
+        internal void Serialize(
+            Utf8JsonWriter writer,
+            in T? rootValue,
+            object? rootValueBoxed = null,
+            bool isInvokedByPolymorphicConverter = false)
```

Comment:
Can the logic that uses these values be refactored (yielding material benefit) into a separate private method since the only calls that pass them are within the type itself?

Reply:
I considered this, but it would end up in a lot of duplication (4x duplication of serialization logic per mode (typed/untyped/polymorphic/non-polymorphic) x (Writer/Sync streaming/Async streaming)). I ended up with this compromise which doesn't sacrifice performance substantially.